controlplane/kubeadm/internal/workload_cluster_etcd.go

```diff
@@ -93,22 +93,24 @@ loopmembers:
 }
 
 // UpdateEtcdVersionInKubeadmConfigMap sets the imageRepository or the imageTag or both in the kubeadm config map.
-func (w *Workload) UpdateEtcdVersionInKubeadmConfigMap(ctx context.Context, imageRepository, imageTag string, version semver.Version) error {
-	return w.updateClusterConfiguration(ctx, func(c *bootstrapv1.ClusterConfiguration) {
+func (w *Workload) UpdateEtcdVersionInKubeadmConfigMap(imageRepository, imageTag string) func(*bootstrapv1.ClusterConfiguration, *[]string) {
```

Comment:
I think we are missing something for etcd.
If I look at what the webhook allows to change, we can change everything both for local etcd and external etcd.

Instead here (and in UpdateEtcdExtraArgsInKubeadmConfigMap) we do not persist changes to:
- etcd.local.DataDir
- etcd.local.ServerCertSANs
- etcd.local.PeerCertSANs
- etcd.External.*

What about refactoring UpdateEtcdVersionInKubeadmConfigMap and UpdateEtcdExtraArgsInKubeadmConfigMap into UpdateEtcdLocalInKubeadmConfigMap (+ filling the gaps) and adding a new UpdateEtcdExternalInKubeadmConfigMap to be called from controlplane/kubeadm/internal/controllers/upgrade.go?

Reply:
Sounds good, we can additionally persist all local and external properties of etcd in the config map.
Since this requires a refactoring to the current methods and thereafter the interface as well (and tests),  it makes sense to combine it with the interface clean-up addressed here IMHO https://github.com/kubernetes-sigs/cluster-api/pull/10154#discussion_r1495041784, wdyt?

Reply:
let's tackle this in this PR (thinking about a potential backport in the 1.6 branch)